### ISSUE TITLE
Computing the mean of a merged layer along one dimension

### ISSUE THREAD

**Issue by kracwarlock** [fchollet/keras#2583]
```
input = ...
left  = LSTM(output_dim=5, return_sequences=True, name='left')(input)
right = LSTM(output_dim=5, return sequences=True, name='right', go_backwards=True)(input)
mix   = merge([left,right], mode='concat')
```
`mix` is of shape `(None,None,10)`
Now I want to take average along the time dimension to get a layer of shape `(None,10)` which takes the mean of all the sequences in `mix`. How can this be done?

**Comment by braingineer** [comment]
note: untested, but something akin to this:

```python
from keras.layers import Lambda
sum_dim1 = Lambda(lambda xin: K.sum(xin, axis=1), output_shape=(None,10))
```

**Comment by kracwarlock** [comment]
This gives me `'Lambda' object has no attribute 'ndim'`

**Comment by braingineer** [comment]
That's a layer. you have to use it. 

```python
sum_dim1(mix)
```

**Comment by kracwarlock** [comment]
Got it. Lambda layer does not support masking. I guess I will have to implement this myself.

**Comment by kracwarlock** [comment]
Thanks.

**Comment by braingineer** [comment]
Ah ya.  if you're dealing with masks you pretty much have to write your own layers.  

Though, as a side note, I did write this one for some quick fix stuff:

```python
class LambdaMask(Layer):
    def __init__(self, func, *args, **kwargs):
        self.func = func
        self.supports_masking = True
        super(LambdaMask, self).__init__(*args, **kwargs)

    def compute_mask(self, x, mask=None):
        return self.func(x, mask)

    def call(self, x, mask=None):
        return x

```

**Comment by sergeyf** [comment]
So is that a bug then? It's as easy as adding the `self.supports_masking = True` to the Keras code?

**Comment by kracwarlock** [comment]
I'll close this.

**Comment by braingineer** [comment]
to @sergeyf  --- not a bug; this just isn't an enterprise product.  It's well formed code, but this is the bleeding edge.   I think it's that no one has written tests to ensure that masking works correctly on a bunch of places.   If you arbitrarily add that, you will get some bizarre circumstances. 

But, the general trace of how Keras compiles its nodes:
1. start with a tensor
2. for each layer it is linked to, get its output via `__call__`
   
   -  note: the logic for this is handled by `keras.engine.topology.Layer` and shouldn't really be overridden, though it is good to read to understand what is going on.

3. for each layer it is linked to, call `compute_mask`

You can explore compiled code by investigating the `__dict__` of the layer/tensor objects.  Keras has a reference trace through everything.  It's either in the object properties `_keras_history`, `outbound_nodes` or `inbound_nodes`.

**Comment by sergeyf** [comment]
Thank you @braingineer that is helpful.
